### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by yiitz** [IBM/etcd-java#20]
hi,I got error log about an hour after my service running.My code:

```
    void main(){

        KvStoreClient etcdClient = EtcdClient.forEndpoints(Arrays.asList((appProperties.getProperty("etcd.endpoints").split(","))))
                .withPlainText()
                .build();
        kvClient = etcdClient.getKvClient();
        leaseClient = etcdClient.getLeaseClient();
        leaseId = leaseClient.create(Long.valueOf(appProperties.getProperty("etcd.ttl"))).get().getID();
        etcdRegisterService();
        ScheduledFuture future = etcdKeepalive();
}

    private void etcdRegisterService() {
        String serviceName = MessageFormat.format("{0}/{1}:{2}",
                appProperties.getProperty("etcd.service.name"),
                appProperties.getProperty("etcd.service.addr"),
                appProperties.getProperty("grpc.port"));
        kvClient.put(ByteString.copyFromUtf8(serviceName),
                ByteString.copyFromUtf8(String.format("{\"Op\":0,\"Addr\":\"%s:%s\",\"Metadata\":{}}",
                        appProperties.getProperty("etcd.service.addr"),
                        appProperties.getProperty("grpc.port"))), leaseId).sync();
        logger.info("register etcd at {}", serviceName);
    }

    private ScheduledFuture etcdKeepalive() {

        TaskScheduler etcdScheduler = (TaskScheduler) applicationContext.getBean("etcdScheduler");
        return etcdScheduler.scheduleAtFixedRate(() -> {
            try {
                leaseClient.keepAliveOnce(leaseId).get();
            } catch (InterruptedException ignored) {
            } catch (Exception e) {
                logger.warn("etcd keepalive error", e);
            }
        }, Duration.ofSeconds(Long.valueOf(appProperties.getProperty("etcd.ttl")) - 3));
    }
```

And the error message:

```
2018-12-24 13:27:55.381 [etcd-event-pool-3] ERROR io.netty.util.ResourceLeakDetector - LEAK: ByteBuf.release() was not called before it's garbage-collected. See http://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:331)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:185)
	io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:121)
	io.grpc.netty.NettyWritableBufferAllocator.allocate(NettyWritableBufferAllocator.java:51)
	io.grpc.internal.MessageFramer.writeKnownLengthUncompressed(MessageFramer.java:226)
	io.grpc.internal.MessageFramer.writeUncompressed(MessageFramer.java:167)
	io.grpc.internal.MessageFramer.writePayload(MessageFramer.java:140)
	io.grpc.internal.AbstractStream.writeMessage(AbstractStream.java:53)
	io.grpc.internal.ForwardingClientStream.writeMessage(ForwardingClientStream.java:37)
	io.grpc.internal.ClientCallImpl.sendMessage(ClientCallImpl.java:422)
	io.grpc.ForwardingClientCall.sendMessage(ForwardingClientCall.java:37)
	io.grpc.ForwardingClientCall.sendMessage(ForwardingClientCall.java:37)
	io.grpc.stub.ClientCalls$CallToStreamObserverAdapter.onNext(ClientCalls.java:335)
	com.ibm.etcd.client.GrpcClient$ResilientBiDiStream$RequestSubStream.sendOnNext(GrpcClient.java:413)
	com.ibm.etcd.client.GrpcClient$ResilientBiDiStream$RequestSubStream.lambda$onNext$0(GrpcClient.java:407)
	com.ibm.etcd.client.SerializingExecutor$TaskRun.run(SerializingExecutor.java:74)
	io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:326)
	io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)

```

lib version: grpc-1.17.1-SNAPSHOT

**Comment by njhill** [comment]
@yiitz I was able to reproduce this, and have included a fix in 0.0.9 which has now been released. Please let me know if it works.

However as mentioned there is built in capability for maintaining a lease automatically which should be used in preference to your approach. This includes a default option to get the id of the client's shared "session lease", e.g. via `etcdClient.getSessionLease().getLeaseId()` which is kept alive until the client is closed or loses connection for longer than the default interval.

Thanks again for the report.

**Comment by njhill** [comment]
@yiitz I'm closing this now since it's likely resolved per the above. Please open again if you still have an issue, thanks!
